### PR TITLE
TimeLock Rollback Documentation

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:

* Formalise how to revert from TimeLock to embedded, in case the AtlasDB team is uncontactable when this needs to be done.
* Codify best practices for reverting, so other/future members of AtlasDB will be more familiar with the process.

**Implementation Description (bullets)**:

* Add a new full page to the docs on Reverse Migrations.

**Concerns (what feedback would you like?)**:

* Are there edge cases I haven't handled here?
* Should I explain the rationale for steps more (or less)?

**Where should we start reviewing?**:

* Anywhere

**Priority (whenever / two weeks / yesterday)**:

* Next week? I want to get this in but it's not absolutely critical.

I would tag James/Saahil for SA but can't seem to find their tags.

### CODE REVIEW

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+Reverse Migration
+=================
+
+.. danger::
+
+   AtlasDB does not natively support reverse migrations. Improperly executing reverse migration from external timestamp
+   and lock services can result in **SEVERE DATA CORRUPTION**! Please contact the AtlasDB team before attempting a
+   reverse migration.
+
+If one wishes to downgrade from an external TimeLock Server to embedded timestamp and lock services, one can perform
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: I find this a bit of a mouthful. How about:

"While the migration from an embedded to external timelock service is automatic, the reverse is a manual process. The steps are the inverse of <Manual Migration to Timelock Server>, with an additional step to update the timestamp bound of the embedded service."

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+the inverse of the database migrations mentioned in :ref:`manual-timelock-migration`. It is also important to update the
+embedded timestamp bound to account for any timestamps issued since the original migration.
+
+The reverse migration process must be run offline (with no AtlasDB clients running) and consists of the following steps.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: Do we need to shut down all clients, or just the one you are migrating? Could also clarify that we don't protect against accidentally not running this offline.

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+Throughout this document, we assume the AtlasDB client you are trying to reverse-migrate is ``client`` .
+
+#. Shut down your AtlasDB clients.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: Should make this clearer that we are either talking about just the one client, or *all* clients (as per above comment)

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+#. Take a restore timestamp.
+#. Update your key-value service to the restore timestamp.
+#. Revert your AtlasDB client configuration.
+#. Shut down the TimeLock servers.
+#. Remove your client from the list of TimeLock clients.
+#. Restart your AtlasDB clients.
+#. Restart the TimeLock servers.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: nit: restart -> start

If I was doing this migration, I'd do this:

1. Stop client
2. Alter timelock configuration
3. Restart timelock
4. Alter client config, start client

My motivation is that I probably have many services depending on timelock, and i don't want to have an unnecessary outage in them

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+Obtain a fresh timestamp for ``client`` from the TimeLock servers. This is important to ensure that the sequence
+of timestamps across TimeLock and the key-value service backed services is monotonically increasing.
+
+This can be obtained using ``curl`` as follows. Note that if you obtain an HTTP 503 error, this is because you curled a
+node that was not the leader; please try the other nodes.
+
+.. code-block:: none
+
+   $ curl localhost:8421/client/timestamp/fresh-timestamp
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: -X POST?

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+   123456789
+
+Note down the value returned from this call; we'll refer to it as ``TS`` from here on.
+
+Step 3: Update your Key-Value Service
+-------------------------------------
+
+Update the timestamp table in your key value service, such that it is valid and has the bound set to ``TS`` .
+This will vary depending on your choice of key-value service.
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: nit, would go with : at the end of the sentence, and bold or otherwise highlight "cassandra", "dbkvs".

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+- If you are rolling back an automated migration with Cassandra or a migration using the steps outlined in
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: Would shorten this to just "if you are using Cassandra", I don't think we need to cover the case of anybody rolling their own DB invalidation

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  :ref:`Manual TimeLock Migration <manual-timelock-migration>`, then you can carry out the following steps.
+  We suppose that your clients have the keyspace ``client`` in the examples below; at a high level, we want to truncate
+  the timestamp table to remove all entries, and then add a single entry with row and column names ``ts`` and a blob
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: nit: ts -> TS?

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+     (0 rows)
+
+     cqlsh:client> INSERT INTO "_timestamp" (key, column1, column2, value) VALUES (0x7473, 0x7473, -1, 0x00000000075bcd15);
+     cqlsh:client> SELECT * FROM "_timestamp" ;
+
+      key    | column1 | column2 | value
+     --------+---------+---------+--------------------
+      0x7473 |  0x7473 |      -1 | 0x00000000075bcd15
+
+  You will need to encode ``TS`` into an 8-byte hex representation; this can be done using the ``printf`` command:
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: printf trick is awesome! Suggest we stick this instruction before the cqlsh command where that hex value is used.

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  it suffices to rename the column back:
+
+  .. code:: sql
+
+     ALTER TABLE atlasdb_timestamp RENAME LEGACY_last_allocated TO last_allocated;
+
+Step 4: Revert AtlasDB Client Configurations
+--------------------------------------------
+
+Change the configuration on your AtlasDB clients to not reference TimeLock any longer. For more detail on options
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: Could be clearer that we're talking about removing the 'timelock' block here.

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+------------------------------------
+
+Restart your AtlasDB clients. At this point, it may be useful to perform a simple smoke test to verify that your
+clients work properly (even while TimeLock is down).
+
+Step 8: Restart your TimeLock Servers
+-------------------------------------
+
+Finally, restart your TimeLock servers. Other services that were still dependent on TimeLock (if any) should now
+work normally. To verify that your client no longer uses TimeLock, it may be useful to curl the fresh-timestamp
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: Extending my comment about the ordering...I'd suggest putting 8 before 7. If there is a weird thing where this namespace isn't gone and your client starts hitting timelock, by the time you've started up timelock, it's too late. I'd prefer to have timelock up and confirming that the namespace doesn't exist before starting my client.

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 12]
```diff
@@ -0,0 +1,140 @@
+.. _timelock-reverse-migration:
+
+Reverse Migration
+=================
+
+.. danger::
+
+   AtlasDB does not natively support reverse migrations. Improperly executing reverse migration from external timestamp
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: "Natively support" seems woolly to me. What does this mean exactly?

My understanding is that we don't want people to run reverse migrations without looping one of us in?

Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: I changed this from the original "does not support" because that wasn't true. I think I'll remove the first sentence altogether.

Yes, your understanding is correct (though we've already said that in the last sentence here).

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 1 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: Okay, agreed - changed wording to match yours (presumably with inverse of the ref).

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 2 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: Clarified that it's only the one you are migrating, and also added a note that there is no protection.

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 3 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: Okay, changed (all clients for that one client).

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 5 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: Yes. Good catch.

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 7 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: Alright, fair enough.

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 8 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: Actually this is the row and column name for the timestamp cell in Cassandra, and **must** be lowercase `ts` (uppercase TS will cause us to start from zero again, I think). I've added a note that this is case sensitive.

Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: On second thoughts, changed to just specify `0x7473` instead, which is what you'll need to enter into cqlsh anyway.

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 9 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: OK. I've reordered this to before - do check out how it looks now, though I have a minor concern that users might try running the printf command from within cqlsh.

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 10 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: Sure, changed to "Remove the timelock block from your AtlasDB client configurations."

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 11 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: I'd say both orderings are safe, actually. 

If you put 7 before 8, if the client was still configured to use timelock then it'll fail to start (Atlas pings it on startup, fails, exits).
If you put 8 before 7, then the NotFoundException should be sufficient (and if a client is still trying to use timelock it'll fail).

But you're right that this minimises downtime; the conditions I did this under was during a maintenance window (so this wasn't important at that time). Changed to put 8 before 7.

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 4 above)
Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: Yep. I've changed the final lineup so that it's closer to this (though not precisely since there is a bit of machinery needed to do the client config right). The final solution has the same amount of downtime I think in that the client is down for the entire upgrade but timelock is down for a single, immediate-ish timelock bounce. Thanks for the suggestion!

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 8 above)
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: Oh gotcha. I was confusing this with the "Note down the value returned from this call; we'll refer to it as `TS` from here on" sentence from earlier.

Perhaps we could alter that earlier sentence to "we'll refer to it as `T` from here on" to avoid confusion?

File: docs/source/services/timelock_service/reverse-migration.rst (hunk 11 above)
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: Definitely in sync that they're both safe. My feelings here (based on my recent maintenance experience!) are that we should be thinking in terms of minimal downtime across the board. Thanks for making the change, glad it looks good to you too :)

File: docs/source/services/timelock_service/reverse-migration.rst [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+     cqlsh:client> TRUNCATE "_timestamp";
+     cqlsh:client> SELECT * FROM "_timestamp" ;
+
+      key | column1 | column2 | value
+     -----+---------+---------+-------
+
+     (0 rows)
+
+  You will need to encode ``TS`` into an 8-byte hex representation; this can be done using the ``printf`` command:
```
Comment (harrybiddle) [inline docs/source/services/timelock_service/reverse-migration.rst]: Agree that users might put this into clqsh, although the `$` is going to help. Can we say "using the `printf` shell command"?

Comment (jeremyk-91) [inline docs/source/services/timelock_service/reverse-migration.rst]: Yep. I've added an explicit mention it's a shell command
